File: android-app/app/src/main/java/lvt/crm/ui/duties/DutyScheduleRange.kt
package lvt.crm.ui.duties

import java.util.Calendar

data class DutyScheduleRange(
    val mode: String,
    val startIso: String,
    val endIso: String,
    val title: String,
) {
    val filename: String
        get() = "LCT tu ${formatDayMonth(startIso).replace("/", ".")}-${formatDayMonth(endIso).replace("/", ".")}.pdf"

    val shortLabel: String
        get() = "${formatVnDate(startIso)} – ${formatVnDate(endIso)}"
}

fun todayIsoDate(calendar: Calendar = Calendar.getInstance()): String = toIsoDate(calendar)

fun scheduleRange(mode: String, anchorIso: String): DutyScheduleRange {
    val anchor = fromIsoDate(anchorIso)
    if (mode == "month") {
        val start = anchor.clone() as Calendar
        start.set(Calendar.DAY_OF_MONTH, 1)
        val end = start.clone() as Calendar
        end.set(Calendar.DAY_OF_MONTH, end.getActualMaximum(Calendar.DAY_OF_MONTH))
        val startIso = toIsoDate(start)
        val endIso = toIsoDate(end)
        return DutyScheduleRange(
            mode = "month",
            startIso = startIso,
            endIso = endIso,
            title = "LỊCH CÔNG TÁC TỪ NGÀY ${formatVnDate(startIso)} ĐẾN ${formatVnDate(endIso)}",
        )
    }
    val start = startOfWeekMonday(anchor)
    val end = start.clone() as Calendar
    end.add(Calendar.DAY_OF_MONTH, 5)
    val startIso = toIsoDate(start)
    val endIso = toIsoDate(end)
    return DutyScheduleRange(
        mode = "week",
        startIso = startIso,
        endIso = endIso,
        title = "LỊCH CÔNG TÁC TỪ NGÀY ${formatVnDate(startIso)} ĐẾN ${formatVnDate(endIso)}",
    )
}

fun shiftScheduleAnchor(mode: String, anchorIso: String, direction: Int): String {
    val anchor = fromIsoDate(anchorIso)
    if (mode == "month") {
        anchor.set(Calendar.DAY_OF_MONTH, 1)
        anchor.add(Calendar.MONTH, direction)
        return toIsoDate(anchor)
    }
    val start = startOfWeekMonday(anchor)
    start.add(Calendar.DAY_OF_MONTH, direction * 7)
    return toIsoDate(start)
}

private fun startOfWeekMonday(date: Calendar): Calendar {
    val start = date.clone() as Calendar
    val day = start.get(Calendar.DAY_OF_WEEK)
    val delta = if (day == Calendar.SUNDAY) -6 else Calendar.MONDAY - day
    start.add(Calendar.DAY_OF_MONTH, delta)
    return start
}

private fun fromIsoDate(value: String): Calendar {
    val parts = value.split("-").map { it.toInt() }
    return Calendar.getInstance().apply {
        clear()
        set(parts[0], parts[1] - 1, parts[2])
    }
}

private fun toIsoDate(date: Calendar): String {
    val year = date.get(Calendar.YEAR)
    val month = date.get(Calendar.MONTH) + 1
    val day = date.get(Calendar.DAY_OF_MONTH)
    return "%04d-%02d-%02d".format(year, month, day)
}

private fun formatDayMonth(isoDate: String): String {
    val parts = isoDate.split("-")
    if (parts.size != 3) return ""
    return "${parts[2]}/${parts[1].toInt()}"
}

private fun formatVnDate(isoDate: String): String {
    val parts = isoDate.split("-")
    if (parts.size != 3) return ""
    return "${parts[2]}/${parts[1].toInt()}/${parts[0]}"
}

File: ios-uikit-lvt/LvtCrmUIKit/DutyScheduleRange.swift
import Foundation

struct DutyScheduleRange: Equatable {
    let mode: String
    let startIso: String
    let endIso: String
    let title: String

    var filename: String {
        "LCT tu \(formatDayMonth(startIso).replacingOccurrences(of: "/", with: "."))-\(formatDayMonth(endIso).replacingOccurrences(of: "/", with: ".")).pdf"
    }

    var shortLabel: String {
        "\(formatVnDate(startIso)) – \(formatVnDate(endIso))"
    }
}

enum DutyScheduleCalendar {
    static func todayIso(_ date: Date = Date(), calendar: Calendar = .current) -> String {
        toIsoDate(date, calendar: calendar)
    }

    static func scheduleRange(mode: String, anchorIso: String, calendar: Calendar = .current) -> DutyScheduleRange {
        let anchor = fromIsoDate(anchorIso, calendar: calendar)
        if mode == "month" {
            let start = calendar.date(from: calendar.dateComponents([.year, .month], from: anchor)) ?? anchor
            let dayCount = calendar.range(of: .day, in: .month, for: start)?.count ?? 1
            let end = calendar.date(byAdding: .day, value: dayCount - 1, to: start) ?? start
            let startIso = toIsoDate(start, calendar: calendar)
            let endIso = toIsoDate(end, calendar: calendar)
            return DutyScheduleRange(
                mode: "month",
                startIso: startIso,
                endIso: endIso,
                title: "LỊCH CÔNG TÁC TỪ NGÀY \(formatVnDate(startIso)) ĐẾN \(formatVnDate(endIso))"
            )
        }
        let start = startOfWeekMonday(anchor, calendar: calendar)
        let end = calendar.date(byAdding: .day, value: 5, to: start) ?? start
        let startIso = toIsoDate(start, calendar: calendar)
        let endIso = toIsoDate(end, calendar: calendar)
        return DutyScheduleRange(
            mode: "week",
            startIso: startIso,
            endIso: endIso,
            title: "LỊCH CÔNG TÁC TỪ NGÀY \(formatVnDate(startIso)) ĐẾN \(formatVnDate(endIso))"
        )
    }

    static func shiftScheduleAnchor(mode: String, anchorIso: String, direction: Int, calendar: Calendar = .current) -> String {
        let anchor = fromIsoDate(anchorIso, calendar: calendar)
        if mode == "month" {
            let start = calendar.date(from: calendar.dateComponents([.year, .month], from: anchor)) ?? anchor
            let shifted = calendar.date(byAdding: .month, value: direction, to: start) ?? start
            return toIsoDate(shifted, calendar: calendar)
        }
        let start = startOfWeekMonday(anchor, calendar: calendar)
        let shifted = calendar.date(byAdding: .day, value: direction * 7, to: start) ?? start
        return toIsoDate(shifted, calendar: calendar)
    }
}

private func startOfWeekMonday(_ date: Date, calendar: Calendar) -> Date {
    let day = calendar.startOfDay(for: date)
    let weekday = calendar.component(.weekday, from: day)
    let delta = weekday == 1 ? -6 : 2 - weekday
    return calendar.date(byAdding: .day, value: delta, to: day) ?? day
}

private func fromIsoDate(_ value: String, calendar: Calendar) -> Date {
    let parts = value.split(separator: "-").compactMap { Int($0) }
    guard parts.count == 3 else { return Date() }
    var components = DateComponents()
    components.year = parts[0]
    components.month = parts[1]
    components.day = parts[2]
    return calendar.date(from: components) ?? Date()
}

private func toIsoDate(_ date: Date, calendar: Calendar) -> String {
    let parts = calendar.dateComponents([.year, .month, .day], from: date)
    return String(format: "%04d-%02d-%02d", parts.year ?? 0, parts.month ?? 0, parts.day ?? 0)
}

private func formatDayMonth(_ isoDate: String) -> String {
    let parts = isoDate.split(separator: "-").map(String.init)
    guard parts.count == 3, let month = Int(parts[1]) else { return "" }
    return "\(parts[2])/\(month)"
}

private func formatVnDate(_ isoDate: String) -> String {
    let parts = isoDate.split(separator: "-").map(String.init)
    guard parts.count == 3, let month = Int(parts[1]) else { return "" }
    return "\(parts[2])/\(month)/\(parts[0])"
}
